File: core/java/src/main/java/hara/lang/protocol/IComponent.java
package hara.lang.protocol;

import hara.lang.declaration.HaraMethod;
import hara.lang.declaration.HaraProtocolBinding;

@HaraProtocolBinding(namespace = "std.protocol.icomponent", name = "IComponent")
public interface IComponent {

  @HaraMethod(value = "props", arity = 1)
  IMetadata getProps();

  @HaraMethod(value = "status", arity = 1)
  IMetadata getStatus();

  /** Implementations may interpret level and return their own query value. */
  @HaraMethod(value = "info", arity = 2)
  default Object info(Object level) {
    return getStatus();
  }

  /** A health result may be structured data, not only a boolean. */
  @HaraMethod(value = "health", arity = 1)
  default Object health() {
    return isStarted();
  }

  @HaraMethod(value = "started?", arity = 1)
  boolean isStarted();

  @HaraMethod(value = "stopped?", arity = 1)
  boolean isStopped();

  @HaraMethod(value = "start", arity = 1)
  IComponent start();

  @HaraMethod(value = "stop", arity = 1)
  IComponent stop();

  @HaraMethod(value = "kill", arity = 1)
  default IComponent kill() {
    return this.stop();
  }

  @HaraMethod(value = "remote?", arity = 1)
  default boolean isRemote() {
    return false;
  }
}

File: core/rust/src/lang/protocol/icomponent.rs
use hara_protocol_macros::hara_protocol;

#[hara_protocol(namespace = "std.protocol.icomponent", name = "IComponent")]
pub trait IComponent {
    type Metadata;
    type QueryLevel;
    type Health;

    #[hara_method(value = "props", arity = 1)]
    fn props(&self) -> Self::Metadata;
    #[hara_method(value = "status", arity = 1)]
    fn status(&self) -> Self::Metadata;
    #[hara_method(value = "info", arity = 2)]
    fn info(&self, _level: &Self::QueryLevel) -> Self::Metadata {
        self.status()
    }
    #[hara_method(value = "health", arity = 1)]
    fn health(&self) -> Self::Health;
    #[hara_method(value = "started?", arity = 1)]
    fn started(&self) -> bool;
    #[hara_method(value = "stopped?", arity = 1)]
    fn stopped(&self) -> bool;
    #[hara_method(value = "start", arity = 1)]
    fn start(&mut self);
    #[hara_method(value = "stop", arity = 1)]
    fn stop(&mut self);
    #[hara_method(value = "kill", arity = 1)]
    fn kill(&mut self) {
        self.stop();
    }
    #[hara_method(value = "remote?", arity = 1)]
    fn remote(&self) -> bool {
        false
    }
}
